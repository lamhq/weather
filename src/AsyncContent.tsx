import React from 'react';
import styled from 'styled-components';
import ScaleLoader from 'react-spinners/ScaleLoader';
import Button from 'react-bootstrap/Button';
import { RetryableError } from './types';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 10px 0;
`;

interface AsyncContentProps {
  isLoading: boolean;
  error?: RetryableError;
  hasData?: boolean;
  children: () => React.ReactElement;
}

const AsyncContent = ({ isLoading, error, hasData, children }: AsyncContentProps): React.ReactElement => {
  if (isLoading) {
    return (
      <Container>
        <ScaleLoader color="#e9ecef" loading={isLoading} />
      </Container>
    );
  }

  if (error) {
    return (
      <Container>
        <p>
          There was an error loading your data:&nbsp;
          {error.message}
        </p>
        {error.retry && <Button onClick={error.retry}>Retry</Button>}
      </Container>
    );
  }

  if (!hasData) {
    return <Container>There&apos;s no data to display</Container>;
  }

  return children();
};

export default AsyncContent;
