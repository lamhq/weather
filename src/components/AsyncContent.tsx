import React from 'react';
import styled from 'styled-components';
import ScaleLoader from 'react-spinners/ScaleLoader';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import { RetryableError } from '../types/common';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 10px 0;
`;

interface AsyncContentProps {
  isLoading: boolean;
  children: () => React.ReactElement;
  error?: RetryableError;
  hasData?: boolean;
}

const AsyncContent = ({
  isLoading,
  error,
  hasData,
  children,
}: AsyncContentProps): React.ReactElement => {
  if (isLoading) {
    return (
      <Container role="alert">
        <ScaleLoader color="#e9ecef" loading={isLoading} />
      </Container>
    );
  }

  if (error) {
    const message = `There was an error when loading your data: ${error.message}. `;
    return (
      <Container>
        <Alert variant="danger">
          {message}
          {error.retry && (
            <Button onClick={error.retry} variant="link" size="sm">
              Retry
            </Button>
          )}
        </Alert>
      </Container>
    );
  }

  if (!hasData) {
    return <Container>There&apos;s no data to display</Container>;
  }

  return children();
};

export default AsyncContent;
