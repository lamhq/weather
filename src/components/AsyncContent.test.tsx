import React from 'react';
import { cleanup, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import AsyncContent from './AsyncContent';

describe('AsyncContent component', () => {
  afterEach(cleanup);

  it('should renders loading indicator', () => {
    const { getByRole } = render(<AsyncContent isLoading>{() => <p>Data</p>}</AsyncContent>);
    const element = getByRole(/alert/i);
    expect(element).toBeInTheDocument();
  });

  it('should renders no data text', () => {
    const { getByText } = render(
      <AsyncContent isLoading={false} hasData={false}>
        {() => <p>Data</p>}
      </AsyncContent>,
    );
    const element = getByText(/no data to display/i);
    expect(element).toBeInTheDocument();
  });

  it('should renders error alert', () => {
    const error = { name: 'sample', message: 'sample error message', retry: jest.fn() };
    const { getByText } = render(
      <AsyncContent isLoading={false} error={error}>
        {() => <p>Data</p>}
      </AsyncContent>,
    );
    const msgElement = getByText(/sample error message/i);
    expect(msgElement).toBeInTheDocument();
    const btnElement = getByText(/Retry/i);
    expect(btnElement).toBeInTheDocument();
  });

  it('should renders content passed in render props', () => {
    const { getByText } = render(
      <AsyncContent isLoading={false} hasData>
        {() => <p>loaded data</p>}
      </AsyncContent>,
    );
    const element = getByText(/loaded data/i);
    expect(element).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    let tree = renderer.create(<AsyncContent isLoading>{() => <p>Data</p>}</AsyncContent>).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer
      .create(
        <AsyncContent isLoading={false} hasData={false}>
          {() => <p>Data</p>}
        </AsyncContent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();

    const error = { name: 'sample', message: 'sample error message', retry: jest.fn() };
    tree = renderer
      .create(
        <AsyncContent isLoading={false} error={error}>
          {() => <p>Data</p>}
        </AsyncContent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer
      .create(
        <AsyncContent isLoading={false} hasData>
          {() => <p>loaded data</p>}
        </AsyncContent>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
