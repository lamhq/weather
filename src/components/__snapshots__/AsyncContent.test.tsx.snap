// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncContent component should match snapshot 1`] = `
<div
  className="sc-AxjAm ilACxQ"
  role="alert"
>
  <div
    className="css-0"
  >
    <div
      className="css-1ehwkh4"
    />
    <div
      className="css-1omaxx1"
    />
    <div
      className="css-343uqu"
    />
    <div
      className="css-1e4pia3"
    />
    <div
      className="css-11xacok"
    />
  </div>
</div>
`;

exports[`AsyncContent component should match snapshot 2`] = `
<div
  className="sc-AxjAm ilACxQ"
>
  There's no data to display
</div>
`;

exports[`AsyncContent component should match snapshot 3`] = `
<div
  className="sc-AxjAm ilACxQ"
>
  <div
    className="fade alert alert-danger show"
    role="alert"
  >
    There was an error when loading your data: sample error message. 
    <button
      className="btn btn-link btn-sm"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      Retry
    </button>
  </div>
</div>
`;

exports[`AsyncContent component should match snapshot 4`] = `
<p>
  loaded data
</p>
`;
